### ISSUE TITLE
Ability to get a separate log file for each session, containing the full detailed command logs (not the level from the logger)

### ISSUE THREAD

**Issue by smarie** [theacodes/nox#409]
On projects with quite a number of nox sessions, it is a bit annoying if several sessions fail but several other succeed, to get the logs of all failed sessions. 

The same will definitely happen when sessions can run in parallel (See https://github.com/theacodes/nox/issues/198 )

In my projects I came up with a quick and dirty prototype where the full log (including stdout and stderr of all commands) of each session is written to an independent file. When the session ends, the file is renamed either as `SUCCESS_...` or `ERROR_...` as shown below:

![image](https://user-images.githubusercontent.com/3236794/112484656-6bdb6900-8d7a-11eb-9452-ccfa22a486a7.png)

This would be an increadibly useful feature in my opinion. To reach this I had to make a change in the `popen` routine in order to use `asyncio`. That way I have the possibility to create an os-independent `tee` of the stdout and stderr streams to both the log file and the general nox logger.

It works pretty well for six of my projects for now.

For example in [this nox file](https://github.com/smarie/python-pytest-cases/blob/main/noxfile.py)

Let me know if you think that such feature could be relevant and if so, how you would think of integrating it in `nox`.
Best, 
Sylvain

**Comment by theacodes** [comment]
Oh I like this idea a lot.

**Comment by smarie** [comment]
To make a step towards resolution, here is an attemps to specify the feature

 - a new option `nox.options.logsdir` with default value `None` meaning `".nox/_runlogs"` (note: maybe `.nox/` comes from another parameter, I need to check). Setting this option to `False` would disable log files. Setting this to a valid path string changes the folder. 
 
 - The folder does not need to exist, it is created automatically if it doesn't. However if it exists, it needs to be empty

 - If this option is enabled, a new `FileHandler` is added to the root nox logger (`logging.getLogger("nox")`). 

 - The stream open by the file handler is shared with `popen` (that is the underlying method used to execute all commands). This method is rewritten (as shown in my repository above) using asyncio in order to "tee" the stdout/stderr *both* to the log file stream, **and** to the string in memory for possible appending to the log (depending on `silent` and the level).
 
Now if you agree with the above, here there is a last point for debate. Either we have this log FileHandler use the same logging level as the root - and in that case only the popen part will be "detailed" in the log, not the nox messages themselves ; *OR* we have this log FileHandler use a different, lower-level logging level so that it is truly independent of the nox logging level. THis seems to be feasible as explained [here](https://stackoverflow.com/a/53418323/7262247)

Thoughts ?

**Comment by zdog234** [comment]
I love all of that :)

**Comment by zdog234** [comment]
> Now if you agree with the above, here there is a last point for debate. Either we have this log FileHandler use the same logging level as the root - and in that case only the popen part will be "detailed" in the log, not the nox messages themselves ; OR we have this log FileHandler use a different, lower-level logging level so that it is truly independent of the nox logging level. This seems to be feasible as explained here

I'm tempted to default to `logging.DEBUG`, but I guess anything is fine as long as long as users *can* easily configure it to go down to `DEBUG`

**Comment by smarie** [comment]
Good point. Then I guess we will have no choice but looking closer at [this](https://stackoverflow.com/a/53418323/7262247).

**Comment by smarie** [comment]
Any other feedback from core devs ? I may take a stab at it in a few weeks, but would love the spec to be quite clear before rushing into anything.
